R/teal_lockfile.R

```diff
@@ -39,50 +39,41 @@ teal_lockfile <- function() {
     }
   }
 
-  if (!(is_in_test() || is_r_cmd_check())) {
-    old_plan <- future::plan()
-    # If there is already a parallel (non-sequential) backend, reuse it.
-    if (inherits(old_plan, "sequential")) {
-      future::plan(future::multisession, workers = 2)
-    }
+  shiny::onStop(function() file.remove(lockfile_path))
 
-    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
-    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
-    logger::log_trace("lockfile creation invoked.")
-  }
+  process <- callr::r_bg( # callr process needs to be assigned to an object so does shiny doesn't kill it on shinyApp
+    func = create_renv_lockfile,
+    args = list(
+      lockfile_path = lockfile_path,
+      opts = options()
+    ),
+    # renv setup is orchestrated by its special S3 objects: renv::settings, renv::config and renv::paths
+    #     `package` parameter include `renv` namespace inside the environment of `func = create_renv_lockfile`
+    package = "renv",
+    # callr process will use environmental variables from the main session (parent process)
+    env = Sys.getenv()
+  )
+
+  logger::log_trace("lockfile creation started.")
+  process
 }
 
-create_renv_lockfile <- function(close = FALSE, lockfile_path = NULL) {
-  checkmate::assert_flag(close)
+create_renv_lockfile <- function(lockfile_path = NULL, opts = NULL) {
   checkmate::assert_string(lockfile_path, na.ok = TRUE)
-  promise <- promises::future_promise({
```

Comment:
I'm sorry if I am undermining the whole callr work but isn't it that easy as adding `seed = NULL` argument?

I'm circling back as I can't find the rationale behind switch to callr in there (please redirect me if it's there). It looks like the new standard is ExtendedTask with future-like objects. This goes completely the opposite way. Wanted to understand the reasons.

Reply:
@pawelru the biggest issue with `future` is that we change the `future::plan` and trying to restore it but we can't promise we will restore it on time. Any plan change by the app developer or module developer is a potential risk and highlighted by `future` [authors that one should avoid setting it](https://future.futureverse.org/articles/future-7-for-package-developers.html#avoid-changing-the-future-strategy).


I give you another example which fails the app. Please do following: 
- run the app on @main. You should get an error. 
- run the app again in the same session. App gets through and returns "42" in the module.
- Then please comment out `teal_lockfile()` in teal and try again in a new session. App works.

Problem we're facing is that `future` loses socket to the jobs in previous plans.

```r
plan(multisession)
v1 %<-% {
  Sys.sleep(1)
  cat("Hello world!\n")
  3.14
}
plan(sequential)
v2 %<-% {
  cat("Hello world!\n")
  Sys.sleep(1)
  3.14
}
v1
# Error: MultisessionFuture (<none>) failed to receiving message from cluster RichSOCKnode #1 (PID 93358 on localhost ‘localhost’). The reason reported was ‘Connection to the worker is corrupt’. 
# Post-mortem diagnostic: No process exists with this PID, i.e. the localhost worker is no longer alive. The socket connection to the worker of MultisessionFuture future (<none>) is lost or corrupted: 
# Connection (connection: index=4, description="NA", class="NA", mode="NA", text="NA", opened="NA", can read="NA", can write="NA", id=1786, raw_id="<pointer: 0x6fa>") is no longer valid. 
# There is currently no registered R connection with that index 4. As an example, this may happen if base::closeAllConnections() have been called, for instance via base::sys.save.image() which in turn is called if the R session (pid 90548) is forced to terminate
v2
```

Feel free to suggest any solution involving `future` which will make it more robust. So far I found `callr::r_bg` solution have not thrown any errors.

<details>
<summary> app </summary>

```r
# library(teal)
pkgload::load_all("~/Documents/nest/teal")
setwd("~/Documents/nest/sample_apps/sample_app_renv_lockfile")


app <- teal::init(
  data = teal_data(obj = data.frame()),
  modules = modules(
    module(
      ui = function(id) {
        ns <- NS(id)
        tagList(
          textOutput(ns("text1")),
          textOutput(ns("text2"))
        )
      },
      server = function(id, data) {
        moduleServer(id, function(input, output, session) {
          output$text1 <- renderPrint({
            future::plan("multicore")
            future::future({
              a <- 7
              b <- 3
              c <- 2
              a * b * c
            })
          })
        })
      }
    )
  )
)

runApp(app)

```

</details>

Reply:
Changing the plan closes/looses other tasks so we are using `promise::then` at the end of the execution of the task so that we close the plan when the task is finished. But starting and closing multiple apps one after another might cause that some tasks did not succeed

Reply:
Thank you for the explanation. I think it all makes sense and it's coming down into the topic of changing the plan. This has to be done more carefully on all parties - app developer, core tool developers (us) and also module developers. I think that the golden rule should be as follows: if the plan has been changed (very easy to detect - see below) - do not change it.

So in other words - we (from the core tool perspective) are going to change the plan if it was not set by ~app developer~ anyone else before. This would have to be documented in a prominent place.

(As an extra extension - we can add an option to better control the settings used. Right now it's hardcoded multisession with 2 workers - maybe one wanted to alter it?)

As a result:
- we will not change the plan set by app developers before init
- module developers cannot change the plan set by either of app developers or us

The latter point is actually true right now. Take the lockfile away - if there are multiple modules and each set its own plan then the app would have a problem. I see your example in this way - take teal away and use base shiny with two modules each trying to set its own, different plan. Module developers also needs to be very careful with this.

See my proposal below:
![image](https://github.com/insightsengineering/teal/assets/12943682/893a544b-ed4b-4ee9-85f5-8aba3ba75a12)

Reply:
(and a little bit of more context behind my ask)
I am having a high hopes for `ExtendedTask`. This was advertised as the way to go for long-running tasks. I think I can name a few in which this would be useful. Therefore I think that the plan setting should be at the {teal} core (inside `init()`; optionally controlled with set of options) and all the modules and teal extensions would be able to use it as they wish.

Reply:
@pawelru all makes sense, but what about the case when a module developer has any `future::plan` in the teal_module code,  then:
- we are setting the `plan("multisession")` for renvlock-process in `teal::init`
- custom `teal_module` is called with `plan("sequential")` which automatically loses renvlock-process

If the answer is "module developer should be more careful" I'd say "module developer has already a lot to learn about `teal` and this would be yet another thing to keep in mind"

Reply:
Yes and no :)
Yes as my answer is unfortunately "module developer should be more careful".
No because this has nothing to do with `teal`. These are all `future` requirements and limitations you linked above. Module should be seen as a function and in [here](https://future.futureverse.org/articles/future-7-for-package-developers.html#avoid-changing-the-future-strategy) we have:

> For reasons like the ones mentioned above, refrain from setting [plan()](https://future.futureverse.org/reference/plan.html) in a function. It is better to leave it to the end-user to decided how they want to parallelize. One reason for this is that we can never know how and in what context our code will run. (...) 

> If you still think it is necessary to set [plan()](https://future.futureverse.org/reference/plan.html), make sure to undo when the function exits, also on errors.

Change "function" to "module", "context" into "app" - you know what I mean here.

See my previous message that this is true even without us touching this at all. I will give you an example. What if you are app developer and me and Marcin are module developers. I'm doing `future::plan` so does Marcin. Then you have a problem. My point is that everyone (incl. module developers) who is doing `plan()` needs to be careful with it and adhere to the docs you linked above. That's completely independent of us touching this.

Reply:
@pawelru 

| What if you are app developer and me and Marcin are module developers. I'm doing future::plan so does Marcin. Then you have a problem.

| you might also call a single module (that runs plan()) multiple times (because why not?). And then even with a one module - you as an app developer - would have a problem.

I think both are not teal-specific problems. They can happen in any shiny app and  we shouldn't complicate this even more in `teal`. I think that `teal` messing with `future` is a mistake and brings unnecessary complications. 

I thing about following scenarios:

1. App developer don't use future and none of teal_modules uses future. In `teal` we are allowed set `plan("multisession")` as it hasn't been changed by app dev. When task is finished (`on.exit`) we set it back to original plan.
2. App developer sets `plan(...)`, teal_modules don't use future.  In `teal` we detect that plan has been changed and we call `teal_lockfile()` on the user-specified-plan. Plan set to "sequential" delays an app `init` because teal app awaits `teal_lockfile()` to finish. Luckily, this delays only `init`, so the next session will be started without lags.
3. App developer don't use future and teal_module uses future. And here are two possible ways for this:
  a. teal_module sets `plan(xyz)` - if we keep the same condition - to change `plan(multisession)` only if app developer didn't set, then we introduce a bug, because `on.exit(...)` is possibly invoked too late as our task runs asynchronously and teal_module with `plan(...)` can be called. Yes, I agree that setting the `plan` by module developer is the same sin as us setting this in `teal::init`, but I think we shouldn't immediately punish module developer for one's wrongdoing. Problem is that we can't detect if any module in the app sets `plan(...)` so we can't even prevent the failure. Failure message could be very cryptic and it will be hard to point exactly which `plan()` call caused a problem.
  b. teal_module can't call a `plan` as generally recommended and app developer sets the plan - then we have scenario (2) 
  
Because of (2), (3a) and (3b) I'm against using `future`. I think lockfile should be created asynchronously always and not being dependent on any app-developer or module-developer decision. Looking at `@main` `ExtendedTask` is not utilized properly so I don't see the point of using this at all (`ExtendedTask` is used just to invoke a task) as invoking future can be done without `ExtendedTask` too.

My general opinion is - why to complicate this with `future` while we can simply call `r_bg` without being worried about (2), (3a) and (3b).

Reply:
After experimenting with `callr` and `r_bg` on this branch I personally believe `callr` is simpler to run and easier to manage and does not mess up with `plan`.  There are also `future` warnings and errors pasted to a console when an app is stopped by the process is still running

Reply:
Even though `ExtendedTask` and `future::future_promises` is the new way to go, I think we've been abusing this solution and not utilizing the whole aspects of the flow. I would advise to go with `callr` solution as I see less issues with hanging and unfinished processes. Happy to meet and discuss next week if you guys feel that this conversation is not over yet

Reply:
Thanks for the comments guys. Let's go with callr. I think that it's quite likely that the future will come back to us but we can revisit this later on (if at all).  For now let's go with callr. Resolving this